Refactoring v1

abczzz13 opened the pull request:
## What?
Cleaning up the first part of the mess I made.

## Why?
Clear and clean code will be easier to maintain, debug and be used by others. 

## How?
Restructured some of the components and made sure all the code can be found in logical places. Additionally removed the flask-login module, as everything will be handled with JWT's. Furthermore, added docstrings and additional comments in various places.

## Testing?
No additional tests have been included, as no new features have been inplemented. However, some tests had to be updated to correspond to the new refactored situation (new auth endpoints, etc.).

All tests pass.

## Anything else?
@jodelajo, with this PR the authentication endpoints will be changed from /api/v1/ to /auth/. We should coordinate when to merge this PR, to make sure the endpoints are also changed in the frontend to minimize any downtime.

Additionally, some of the query parameters of the /api/v1/companies endpoint will get a name change to simplify the code in the backend.

Below you can find the overview of the new endpoints:
- /api/v1/create_token -> /auth/create_token
- /api/v1/check_token -> /auth/check_token
- /api/v1/register -> /auth/register

Changes to to the /api/v1/companies parameters:
- ?company=... -> ?company_name=...
- ?size=... -> ?company_size=...
- ?city=... -> ?city_name=...
- ?branch=... -> ?branches=...
- ?discipline=... -> ?disciplines=...
- ?tag=... -> ?tags=...

Reinoptland commented on app/api/companies.py:
```diff
@@ -1,189 +1,77 @@
-                                 param_dict[parameter])
-        if parameter == 'company_like':
+    # Iterate over the query parameters and adjust query accordingly
+    for key in request_dict:
```
Nice, this abstracts away the if/elsing of all the keys. 

Does each query.filter generate a database call though? Or does it execute any call to the db at the end? 🤔

Worth checking out.

abczzz13 replied:
In every step a new queryset will be generated, which will be only executed at the last step (for example when .first(), all() or queryset[0] is requested).

At least, that should be the theory, I will check the calls on the DB to be sure.

Reinoptland commented on app/api/companies.py:
```diff
@@ -1,189 +1,77 @@
+        if key not in parameters:
+            return error_response(400, """The parameter(s) you have used are unknown. \
+                Please use one or multiple of the following parameters: \
+                    {0}, {1}, {2}, {3}, {4}, {5}, {6}, {7}, {8}, {9}, {10}, {11}, {12}""".format(*parameters))
```
In this string the amount of parameters are hardcoded, so this is bound to break quickly especially now that you've abstracted to `key in request_dict:`

abczzz13 replied:
True, it smells and should be corrected!

Reinoptland commented on app/auth/routes.py:
```diff
@@ -1,37 +1,86 @@
-from flask_login import current_user, login_user, logout_user
-from werkzeug.urls import url_parse
+"""
+This module provides all the endpoints for cities API:
```
..? This is Auth right?

Reinoptland replied:
`"""This module provides all the endpoints for authentication API:"""`

abczzz13 replied:
Indeed it is, this will be corrected.

Reinoptland commented on app/auth/routes.py:
```diff
@@ -1,37 +1,86 @@
+@bp.route('token', methods=['DELETE'])
+@jwt_required()
+def revoke_token():
+    """DELETE  /token          Revokes token (not implemented yet) """
```
JWT's are challenging to revoke. What you can do is cross reference some part of the payload of the JWT with the database. For example: 

- Your token has a user_id as a payload
- In the database you have a column `account_blocked (boolean)` 
- When a user makes a request you can check if the token is valid *and* if the account_blocked for that user is true
- This way you can "revoke" **access** eventhough the token is still valid. 
- But checking the db this way on each request is rather costly, which is why you might implement this kind of behaviour for a long lived refresh_token only and accept that the short-lived tokens are not cross referenced on each request. 
- For particular sensitive decisions ( e.g. deleting an account or payments ) you can require a new refresh_token to make sure there is no funny business

And implementing this kind of things is indeed .. a lot of work haha. And just implementing this endpoint will not suffice, it will change the auth system from the ground up

abczzz13 replied:
Yeah, that was kind of what I expected. 

Well, as long as this feature is not requested, I will remove it for now.

Reinoptland commented on app/errors/handlers.py:
```diff
@@ -1,15 +1,54 @@
-from flask import render_template
+"""
+This module provides all the error responses
```
I love this as it standardises the errors (which you often want to display in the frontend) so it's nice if the errors are in predictable format 🙌

Reinoptland commented on app/models.py:
```diff
@@ -1,16 +1,18 @@
+'''Models for the different Tables in DB with additional methods and validation'''
```
I think that eventually models should be a folder, with each model in its own file.

For me personally, when a file reaches 200 lines I start to consider: is this file responsible for too many things?

abczzz13 replied:
Good point, definitely worth considering for the next refactoring (I'd say this PR is already way too large as it is).

Reinoptland commented on app/models.py:
```diff
@@ -186,34 +240,46 @@ def get_or_create(self, meta_string, type):
+    email = ma.Email(
+        validate=validate.Length(min=2, max=64),
+        required=True)
+    password = ma.Str(
```
🤔 are the passwords stored as strings in the database? 

If so: we will need to **hash** the passwords

Let's discuss this

abczzz13 replied:
As discussed, the paswords are hashed and salted by Flask's Werkzeug libary.

Reinoptland reviewed (approved):
I think it's great that you've taken the time to clean up! Rewriting is the essence of software development, so practicing cleaning up is great. I like the restructuring in general, sometimes minor renaming makes it easier to abstract a lot of things as you don't need to translate data formats as much. 

A few thoughts:

- Keep in mind that some kind of versioning in the api might be useful (even if you're dropping the api/v1/ part of the url now. But having the same urls for each version also has it's advantages - [read this](https://restfulapi.net/versioning/) for alternate approaches

- When it comes to the formatting (spacing etc.) I recommend using [prettier](https://prettier.io/docs/en/why-prettier.html), click the link for a few good reasons why,  [here is how to set it up for python](https://dev.to/eegli/quick-guide-to-python-formatting-in-vs-code-2040). 

- Throwing out unused files (like the templates) is great, because that way you don't accumulate cruft over time. This is easier to do with a decent test coverage, as you will notice when you delete something you shouldn't. Detecting unused code in python is challenging as it is not built into the language itself (unlike Java for instance). Doing it regularly as part of each feature is usually best. 

- Putting `@Todo` is nice for the readers of this project because it shows awareness of what's still missing. In real life all `@Todo's` belong in the backlog (Trello, Jira, issues or whatever) as `@Todo's` get lost, become irrelevant. If you must include an `@Todo` for clarity, make a ticket in Trello / Jira as well and link to it in your` @Todo` (the source of truth)

- Did you also update the documentation website? (the swagger thing?) I couldn't find that here, but maybe I missed it

abczzz13 commented:
> 

All the company and city endpoints still use the /api/v1/ prefix. Only the authentication endpoints moved to the /auth/ prefix. Would you say it is also a good idea the use versioning on the /auth/ endpoints?

Moved the Todo: comments to the Trello board and removed the comments.

Documentation on Swagger is also up-to-date with the current state of this PR
